### PR TITLE
Add GAN federated learning support

### PR DESCRIPTION
**Pull request by cuiboyuan**
Add implementations to support training GAN networks with federated averaging on Plato

## Description

To support GAN, new model, trainer, algorithm, and server are added.

### DCGAN Model
For the new model `plato/models/dcgan.py`, different from the traditional models, the GAN model contains two neural networks: a Generator and a Discriminator. For `dcgan.py`, the `Model` class is no longer `nn.Module`. It is just a wrapper class of the Generator and Discriminator class. Users should access attribute `self.generator` and `self.discriminator`of the `Model` class for training or testing purposes. The structure of the DCGAN model is borrowed from this link: https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html 

### GAN Trainer
For the new trainer `plato/trainers/gan.py`, during the training, the label of the data is discarded since GAN does not concern the category of the input data. For GAN, the training is all about making the Generator network generate data realistic enough such that it can fool the Discriminator, while on the other hand making the Discriminator smart enough to be able to distinguish fake data from the real data. If everything goes well, at the end, we will obtain a Generator that can produce data realistic enough to fool human beings. The implementation to train GAN is also borrowed from this link:  https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html 

For the testing of GAN, we are concerned about how realistic the generated data are. Normally, this process is done by human inspections, or by adding the synthetic data to the training set of a classifier network and checking if the accuracy changes (i.e. data augmentation). Both methods are not applicable here. In Plato, we calculate the test accuracy at the end of each round, so we cannot add human interventions. The data augmentation method is "implementable", but it will take a tremendously large amount of time to train a new network, and we have to do it at the end of each round. It is a huge waste of time just to evaluate GAN. We need some method that can finish in a reasonable amount of time. Additionally, both methods do not use test sets at all, which is also a waste of data

For the GAN trainer in Plato, we use a metric called the Frechet Inception Distance (FID). Frechet Distance is a metric to calculate how different two Normal distributions are. The formula is shown below:

039a67f367e74263a3f9e3238f574a7080c430e6

Basically, we assume that real and fake data follow two different multivariate Normal distributions, and we want to bring these two distributions as close as possible. A lower FID indicates a better Generator model.
  
For GAN that generates images, we normally do not assume that images are directly drawn from a Normal distribution. Instead, we assume that there exist some features (sometimes called embeddings) embedded in the image, and that feature follows a Normal distribution. To extract the feature from images, we will use the InceptionV3 model that is pre-trained on ImageNet. The output feature size of InceptionV3 is 2048.

To summarize, the procedure to evaluate the Generator model is: first, loop over the test set and extract the feature of each image with InceptionV3, which represents the distribution of real data; next, generate the same number of fake images with the Generator and extract the features with InceptionV3, which represents the distribution of fake data; finally, with features from both real and fake images, calculate the FID. The implementation to compute FID is borrowed from here: https://wandb.ai/ayush-thakur/gan-evaluation/reports/How-to-Evaluate-GANs-using-Frechet-Inception-Distance-FID---Vmlldzo0MTAxOTI

The FID is not a perfect metric to evaluate GAN and there are several drawbacks. First, FID works the best when there is a large number of samples (typically 50K). Second, FID assumes that these image features follow a multivariate Normal distribution, which sometimes might not be the case. Third, for FID, the features of images are extracted by InceptionV3 pre-trained on ImageNet, which is known to be effective for natural images, but if the input data are not natural images (e.g. hand-written digits, audio signals), the features might not be extracted properly.

Despite all the drawbacks, FID is still the most common metric used to evaluate GAN. We could improve the GAN evaluation in Plato when there are better metrics coming up in the future.

### GAN Federated Averaging Algorithm

For the new algorithm `plato/algorithms/fedavg_gan.py`, it needs to apply the federated averaging on both Generator and Discriminator and load the weights correspondingly. The payload received should be a tuple of `OrderedDict`, which represents the weight of Generator and Discriminator models. The procedure to compute average weights remains the same as the original `fedavg.py` except that we need to do the weight averaging twice, one for Generator and one for Discriminator.

### GAN Federated Averaging Server

For the new server `plato/servers/fedavg_gan.py`, it is almost identical to the original `fedavg.py` except that it needs to update the weight of the Generator and Discriminator separately. Another difference is that for `fedavg_gan.py`, the server does not need to send the weights of both models back to the clients. The server can send both, one of them, or none. This option is configurable in the YAML file. Users just have to specify the following:
```
server:
    address: 127.0.0.1
    port: 8000

    # Choose to send Generator or Discriminator to clients at
    # the end of each round
    network_to_sync: <one of 'none', 'both', 'discriminator', 'generator'>
```
By default, if `network_to_sync` is not in the YAML file, the server will send the weights of both models back to the clients. This idea is borrowed from this paper: https://deepai.org/publication/federated-generative-adversarial-learning 

## How has this been tested?

Execute the following command without error:
```
python3 run -c configs/CelebA/fedavg_gan_dcgan.yml
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: configs/CelebA/fedavg_gan_dcgan.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+server:
+    address: 127.0.0.1
+    port: 8000
+
+    # Choose to send Generator or Discriminator to clients at
+    # the end of each round
+    network_to_sync:
+        generator: true
+        discriminator: false
```
Comment (baochunli) [inline configs/CelebA/fedavg_gan_dcgan.yml:22]: It may be a better idea to just say: `network_to_sync: generator` or `network_to_sync: discriminator`.

Comment (cuiboyuan) [reply configs/CelebA/fedavg_gan_dcgan.yml]: Now updated in the latest commit
